src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java

```diff
@@ -0,0 +1,157 @@
+		}
+		// Try the sets static factory method
+		return getSetsFromStaticFactory(testMethod, annotation.factory());
+	}
+
+	private static void ensureNoInputConflicts(CartesianProductTest annotation,
+			List<CartesianValueSource> valueSources) {
+		boolean hasValue = annotation.value().length != 0;
+		boolean hasFactory = !annotation.factory().isEmpty();
+		boolean hasValueSources = !valueSources.isEmpty();
+		if (hasValue && hasFactory || hasValue && hasValueSources || hasFactory && hasValueSources) {
+			throw new ExtensionConfigurationException(
+				"CartesianProductTest can only take exactly one type of arguments source");
+		}
+	}
+
+	private List<List<?>> getSetsFromValue(Method testMethod, CartesianProductTest annotation) {
+		List<List<?>> sets = new ArrayList<>();
+		List<String> strings = Arrays.stream(annotation.value()).distinct().collect(toList());
+		for (int i = 0; i < testMethod.getParameterTypes().length; i++) {
+			sets.add(strings);
+		}
+		return sets;
+	}
+
+	private List<List<?>> getSetsFromRepeatableAnnotation(List<CartesianValueSource> valueSources) {
+		List<List<?>> sets = new ArrayList<>();
+		for (CartesianValueSource source : valueSources) {
+			CartesianValueArgumentsProvider provider = new CartesianValueArgumentsProvider();
+			provider.accept(source);
+			List<Object> collect = provider.provideArguments().distinct().collect(toList());
+			sets.add(collect);
+		}
+		return sets;
+	}
+
+	private List<List<?>> getSetsFromStaticFactory(Method testMethod, String explicitFactoryName) {
+		if (explicitFactoryName.isEmpty())
+			return invokeSetsFactory(testMethod, testMethod.getName()).getSets();
+		else
+			return invokeSetsFactory(testMethod, explicitFactoryName).getSets();
+	}
+
+	private CartesianProductTest.Sets invokeSetsFactory(Method testMethod, String factoryMethodName) {
+		Class<?> declaringClass = testMethod.getDeclaringClass();
+		Method factory = PioneerUtils
+				.findMethodCurrentOrEnclosing(declaringClass, factoryMethodName)
+				.orElseThrow(() -> new ExtensionConfigurationException("Method `CartesianProductTest.Sets "
+						+ factoryMethodName + "()` not found in " + declaringClass + "or any enclosing class"));
+		String method = "Method `" + factory + "`";
+		if (!Modifier.isStatic(factory.getModifiers())) {
+			throw new ExtensionConfigurationException(method + " must be static");
+		}
+		if (!CartesianProductTest.Sets.class.isAssignableFrom(factory.getReturnType())) {
+			throw new ExtensionConfigurationException(method + " must return `CartesianProductTest.Sets`");
+		}
+		CartesianProductTest.Sets sets = (CartesianProductTest.Sets) invokeMethod(factory, null);
+		if (sets.getSets().size() > testMethod.getParameterCount()) {
+			throw new ParameterResolutionException(format(
+				"%s must register values for each parameter exactly once. Expected [%d] parameter sets, but got [%d]",
+				method, testMethod.getParameterCount(), sets.getSets().size()));
+		}
```

Comment:
What if `sets.getSets().size() == testMethod.getParameterCount()`, but one of the parameters is resolved by JUnit itself? Then we should still see the exception.

Reply:
Can't, JUnit throws an exception before we get here - competing ParameterResolvers.

Reply:
Oh, good to know! Can you add a comment saying as much?